server/src/test/java/org/opensearch/tasks/consumer/TopNSearchTasksLoggerTests.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks.consumer;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LogEvent;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.opensearch.action.search.SearchShardTask;
+import org.opensearch.common.logging.Loggers;
+import org.opensearch.common.logging.MockAppender;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.tasks.ResourceStats;
+import org.opensearch.tasks.ResourceStatsType;
+import org.opensearch.tasks.ResourceUsageMetric;
+import org.opensearch.tasks.Task;
+import org.opensearch.test.OpenSearchSingleNodeTestCase;
+
+import java.util.Collections;
+
+import static org.opensearch.tasks.consumer.TopNSearchTasksLogger.LOG_TOP_QUERIES_FREQUENCY;
+import static org.opensearch.tasks.consumer.TopNSearchTasksLogger.LOG_TOP_QUERIES_SIZE;
+
+public class TopNSearchTasksLoggerTests extends OpenSearchSingleNodeTestCase {
+    static MockAppender appender;
+    static Logger searchLogger = LogManager.getLogger(TopNSearchTasksLogger.TASK_DETAILS_LOG_PREFIX + ".search");
+
+    private TopNSearchTasksLogger topNSearchTasksLogger;
+
+    @BeforeClass
+    public static void init() throws IllegalAccessException {
+        appender = new MockAppender("trace_appender");
+        appender.start();
+        Loggers.addAppender(searchLogger, appender);
+    }
+
+    @AfterClass
+    public static void cleanup() {
+        appender.stop();
+        Loggers.removeAppender(searchLogger, appender);
+    }
+
+    public void testLoggerWithTasks() {
+        final Settings settings = Settings.builder().put(LOG_TOP_QUERIES_SIZE, 1).put(LOG_TOP_QUERIES_FREQUENCY, 10).build();
+        topNSearchTasksLogger = new TopNSearchTasksLogger(settings);
+
+        generateTasks(5);
+        // sleeping enough time for the logger to complete publishing entries to log file
+        try {
+            Thread.sleep(1_000);
+        } catch (InterruptedException e) {
+            // ignore
+        }
+        generateTasks(1);
+        LogEvent logEvent = appender.getLastEventAndReset();
+        assertNotNull(logEvent);
+        assertEquals(logEvent.getLevel(), Level.INFO);
+        assertTrue(logEvent.getMessage().getFormattedMessage().contains("cpu_time_in_nanos=400, memory_in_bytes=400"));
```

Comment:
What is the intent of this test? I might be reading this wrong, but I think you're asserting that the last entry of `generateTasks(5)` is logged but the task from `generateTasks(1)` is not logged? That looks like a built-in race condition. What are you attempting to test here that wouldn't be covered by a simple test of using a frequency setting of zero, generating a single task, and verifying it gets logged?

Reply:
Updated the test with setting of zero and verifying that it is getting logged.